### PR TITLE
util, executor: deep clone shallowRow in baseJoiner.Clone

### LINKED ISSUE #16586: IndexHashJoin with semi join on TPCH dataset returns wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
1. tpch with scale factor = 1

2. get unstable result
``` sql
MySQL [tpch]>  select count(*) from supplier, lineitem l1, orders, nation where s_suppkey = l1.l_suppkey and o_orderkey = l1.l_orderkey and o_orderstatus = 'F' and l1.l_receiptdate > l1.l_commitdate and exists ( select * from lineitem l2 where
 l2.l_orderkey = l1.l_orderkey and l2.l_suppkey <> l1.l_suppkey ) and s_nationkey = n_nationkey and n_name = 'EGYPT';
+----------+
| count(*) |
+----------+
|    73600 |
+----------+
1 row in set (3.32 sec)

MySQL [tpch]>  select count(*) from supplier, lineitem l1, orders, nation where s_suppkey = l1.l_suppkey and o_orderkey = l1.l_orderkey and o_orderstatus = 'F' and l1.l_receiptdate > l1.l_commitdate and exists ( select * from lineitem l2 where
 l2.l_orderkey = l1.l_orderkey and l2.l_suppkey <> l1.l_suppkey ) and s_nationkey = n_nationkey and n_name = 'EGYPT';
+----------+
| count(*) |
+----------+
|    73586 |
+----------+
1 row in set (3.37 sec)
```

3. explain result
``` sql
MySQL [tpch]> explain select count(*) from supplier, lineitem l1, orders, nation where s_suppkey = l1.l_suppkey and o_orderkey = l1.l_orderkey and o_orderstatus = 'F' and l1.l_receiptdate > l1.l_commitdate and exists ( select * from lineitem l2 where l2.l_orderkey = l1.l_orderkey and l2.l_suppkey <> l1.l_suppkey ) and s_nationkey = n_nationkey and n_name = 'EGYPT';
+--------------------------------------------+------------+-----------+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                         | estRows    | task      | access object                                     | operator info                                                                                                                                                                                                                  |
+--------------------------------------------+------------+-----------+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| HashAgg_21                                 | 1.00       | root      |                                                   | funcs:count(1)->Column#55                                                                                                                                                                                                      |
| └─IndexHashJoin_36                         | 192038.88  | root      |                                                   | semi join, inner:IndexLookUp_27, outer key:tpch.lineitem.l_orderkey, inner key:tpch.lineitem.l_orderkey, other cond:ne(tpch.lineitem.l_suppkey, tpch.lineitem.l_suppkey), ne(tpch.lineitem.l_suppkey, tpch.supplier.s_suppkey) |
|   ├─IndexMergeJoin_59(Build)               | 240048.60  | root      |                                                   | inner join, inner:TableReader_57, outer key:tpch.lineitem.l_orderkey, inner key:tpch.orders.o_orderkey                                                                                                                         |
|   │ ├─HashJoin_65(Build)                   | 240048.60  | root      |                                                   | inner join, equal:[eq(tpch.supplier.s_suppkey, tpch.lineitem.l_suppkey)]                                                                                                                                                       |
|   │ │ ├─HashJoin_78(Build)                 | 400.00     | root      |                                                   | inner join, equal:[eq(tpch.nation.n_nationkey, tpch.supplier.s_nationkey)]                                                                                                                                                     |
|   │ │ │ ├─TableReader_83(Build)            | 1.00       | root      |                                                   | data:Selection_82                                                                                                                                                                                                              |
|   │ │ │ │ └─Selection_82                   | 1.00       | cop[tikv] |                                                   | eq(tpch.nation.n_name, "EGYPT")                                                                                                                                                                                                |
|   │ │ │ │   └─TableFullScan_81             | 25.00      | cop[tikv] | table:nation                                      | keep order:false                                                                                                                                                                                                               |
|   │ │ │ └─TableReader_80(Probe)            | 10000.00   | root      |                                                   | data:TableFullScan_79                                                                                                                                                                                                          |
|   │ │ │   └─TableFullScan_79               | 10000.00   | cop[tikv] | table:supplier                                    | keep order:false                                                                                                                                                                                                               |
|   │ │ └─TableReader_86(Probe)              | 4800972.00 | root      |                                                   | data:Selection_85                                                                                                                                                                                                              |
|   │ │   └─Selection_85                     | 4800972.00 | cop[tikv] |                                                   | gt(tpch.lineitem.l_receiptdate, tpch.lineitem.l_commitdate)                                                                                                                                                                    |
|   │ │     └─TableFullScan_84               | 6001215.00 | cop[tikv] | table:l1                                          | keep order:false                                                                                                                                                                                                               |
|   │ └─TableReader_57(Probe)                | 0.49       | root      |                                                   | data:Selection_56                                                                                                                                                                                                              |
|   │   └─Selection_56                       | 0.49       | cop[tikv] |                                                   | eq(tpch.orders.o_orderstatus, "F")                                                                                                                                                                                             |
|   │     └─TableRangeScan_55                | 1.00       | cop[tikv] | table:orders                                      | range: decided by [tpch.lineitem.l_orderkey], keep order:true                                                                                                                                                                  |
|   └─IndexLookUp_27(Probe)                  | 4.03       | root      |                                                   |                                                                                                                                                                                                                                |
|     ├─IndexRangeScan_25(Build)             | 4.03       | cop[tikv] | table:l2, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | range: decided by [eq(tpch.lineitem.l_orderkey, tpch.lineitem.l_orderkey)], keep order:false                                                                                                                                   |
|     └─TableRowIDScan_26(Probe)             | 4.03       | cop[tikv] | table:l2                                          | keep order:false                                                                                                                                                                                                               |
+--------------------------------------------+------------+-----------+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### 2. What did you expect to see?
correct stable result

### 3. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
master

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #16586 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Deep clone shallowRow in baseJoiner.Clone

How it Works:
joiner.shallowRow will not be modified unexpectedly when parallelly execution.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 
Deep clone shallowRow in baseJoiner.Clone to fix a correctness issue when IndexHashJoin execute parallelly

### CODE REVIEW

**Comment by zz-jason** [comment]
does other release branches have this issue?

**Comment by XuHuaiyu** [comment]
@zz-jason The `clone` function is introduced in release-4.0

**Comment by zz-jason** [comment]
> @zz-jason The `clone` function is introduced in release-4.0

Does it mean that release 3.0, 3.1 do not have this bug?

**Comment by XuHuaiyu** [comment]
> Does it mean that release 3.0, 3.1 do not have this bug?

yes

File: executor/join_test.go [hunk 1]
```diff
@@ -1290,8 +1290,64 @@ func (s *testSuiteJoinSerial) TestIndexNestedLoopHashJoin(c *C) {
...
+	tk.MustExec("insert into lineitem values(4,4,2,2)")
+	tk.MustExec("insert into supplier values(4, 4)")
+	tk.MustExec("insert into orders values(4, 'F')")
+	tk.MustExec("insert into nation values(4, 'EGYPT')")
+	tk.MustQuery("select count(*) from supplier, lineitem l1, orders, nation where s_suppkey = l1.l_suppkey and o_orderkey = l1.l_orderkey and o_orderstatus = 'F' and  exists ( select * from lineitem l2 where l2.l_orderkey = l1.l_orderkey and l2.l_suppkey < l1.l_suppkey ) and s_nationkey = n_nationkey and n_name = 'EGYPT' order by l1.l_orderkey, l1.l_linenumber;").Check(testkit.Rows("18"))
+	tk.MustExec("drop table lineitem")
+	tk.MustExec("drop table nation")
+	tk.MustExec("drop table supplier")
+	tk.MustExec("drop table orders")
+	c.Assert(failpoint.Disable("github.com/pingcap/tidb/planner/core/MockOnlyEnableIndexHashJoin"), IsNil)
```
Comment (SunRunAway) [inline executor/join_test.go:1350]: nit: How about using `defer` for this statement?

File: util/chunk/mutrow.go [hunk 2]
```diff
@@ -38,6 +38,18 @@ func (mr MutRow) Len() int {
 	return len(mr.c.columns)
 }
 
+// Clone deep clone a MutRow.
+func (mr MutRow) Clone() MutRow {
```
Comment (SunRunAway) [inline util/chunk/mutrow.go:42]: Is it better to call [`Row.CopyConstruct`](https://github.com/pingcap/tidb/blob/master/util/chunk/row.go#L211) here?

Comment (XuHuaiyu) [reply util/chunk/mutrow.go:42]: If mr.c == nil, Row.CopyConstruct will panic.

**Pull request merged**